### PR TITLE
fix(vitest): fix file snapshots in skipped suites considered obsolete

### PR DESCRIPTION
**Pull request by hi-ogawa**
### Description

Closes https://github.com/vitest-dev/vitest/issues/2442
Supersedes and closes https://github.com/vitest-dev/vitest/pull/4009

Current behavior was quite tricky, so I created (hopefully) cleaner repro here (and this is the base of the test I added in this PR):
- https://stackblitz.com/edit/vitest-dev-vitest-xqw54j?file=repro.test.ts

The repro shows that treating snapshots as "obsolete" is not only about wrong warning, but also it has a consequence of deleting those skipped snapshots when running with `-u`. 

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: packages/vitest/src/runtime/runners/test.ts [hunk 1]
```diff
@@ -27,7 +27,17 @@ export class VitestTestRunner implements VitestRunner {
...
+  async onAfterRunFiles(files?: File[]) {
+    // mark snapshots in skipped tests as not obsolete
+    // TODO: this probably doesn't work when `VitestTestRunner` is handling multiple files concurrently,
+    //       but `snapshotClient.startCurrentRun/finishCurrentRun` doesn't work in that case already?
+    for (const test of getTests(files ?? [])) {
+      if (test.mode === 'skip') {
+        const name = getNames(test).slice(1).join(' > ')
+        this.snapshotClient.skipTestSnapshots(name)
+      }
+    }
```
Comment (hi-ogawa) [inline packages/vitest/src/runtime/runners/test.ts:39]: I'm assuming snapshot system currently doesn't work when there are multiple `files` running here, but I'm not able to reproduce such case yet.
My assumption might be totally wrong, but I'm investigating a little more.

Comment (hi-ogawa) [reply packages/vitest/src/runtime/runners/test.ts:39]: The code should be moved to `onAfterRunSuite` if `SnapshotClient.finishCurrentRun` is moved to `onAfterRunSuite` by https://github.com/vitest-dev/vitest/pull/4796.

File: test/snapshots/test/fixtures/skip-test/repro.test.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  })
+})
+
+it.skipIf(ENABLE_SKIP)('top-level case', () => {
+  expect('hi-2').toMatchSnapshot()
+})
+
+// requires at least one non-skipped test to trigger
+// `SnapshotClient.startCurrentRun` on current file
+it('normal case', () => {
```
Comment (sheremet-va) [inline test/snapshots/test/fixtures/skip-test/repro.test.ts:17]: This looks like an unfortunate edge case. I think this PR should also cover files that have only a single `describe.skipIf`

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: test/snapshots/test/fixtures/skip-test/repro.test.ts (hunk 2 above)
Comment (hi-ogawa) [reply test/snapshots/test/fixtures/skip-test/repro.test.ts:17]: > I think this PR should also cover files that have only a single `describe.skipIf`

Sorry, the comment here might be confusing, but the reason why I have this is to reproduce the bug. Without this "normal case", the old code looks working and new code is also working (so this is for 2nd checkbox, "include a test that fails without this PR but passes with it.")

Does this make sense? Or did mean something else?

Comment (sheremet-va) [reply test/snapshots/test/fixtures/skip-test/repro.test.ts:17]: Do you mean that without this `it`, the code works before and after? But with this `it` the code didn't work before?

Comment (hi-ogawa) [reply test/snapshots/test/fixtures/skip-test/repro.test.ts:18]: Yes, that's what I mean. I adjusted my comment a bit https://github.com/vitest-dev/vitest/pull/4795/commits/4b480ef0f0c73b257dcc3c442e3579a94dcb4d23

Comment (sheremet-va) [reply test/snapshots/test/fixtures/skip-test/repro.test.ts:18]: Ok then 👍🏻

**Review by sheremet-va — APPROVED** [review]
